src/Agent/NewRelic/Agent/Core/Samplers/GcSampler.cs

```diff
@@ -61,186 +64,304 @@ public class GcSampler : AbstractSampler
+        {
+            if (!Log.IsEnabledFor(level))
+            {
+                return;
+            }
+
+            if (ex == null)
+            {
+                Log.LogMessage(level, $"GC Performance Counters: {message}");
+            }
+            else
+            {
+                Log.LogMessage(level, $"GC Performance Counters: {message}, {ex}");
+            }
+        }
+
+        /// <summary>
+        /// Method ensures that the performance counters are available and up-to-date.
+        /// If the Performance Counter Instance name has changed since last sampling, create new performance
+        /// counters.
+        /// </summary>
+        /// <returns>bool indicating success of doing so</returns>
+        private bool EnsurePerformanceCounters(string currentProcessInstanceName)
+        {
+            if (currentProcessInstanceName == _perfCounterInstanceName && _perfCounterProxies != null && _perfCounterProxies.Count > 0)
+            {
+                return true;
+            }
+
+            if (!string.IsNullOrWhiteSpace(_perfCounterInstanceName))
+            {
+                LogMessage(LogLevel.Debug, $"Performance Counter Instance name change detected, rebuilding performance counters:  changed from '{_perfCounterInstanceName}' to '{currentProcessInstanceName}'");
+            }
+
+            _perfCounterInstanceName = currentProcessInstanceName;
+
+            var countPerfCountersCreated = CreatePerfCounterProxies(_perfCounterInstanceName);
+
+            return countPerfCountersCreated > 0;
+        }
+
+        private void HandleUnauthorizedException()
+        {
+            var userName = "<unknown>";
+            try
+            {
+                userName = System.Security.Principal.WindowsIdentity.GetCurrent().Name;
+            }
+            catch
+            { }
+
+            LogMessage(LogLevel.Warn, $"The executing user, {userName}, has insufficient permissions to collect Windows Performance Counters.");
+
+            Stop();
+        }
+
+        private void HandleException(Exception ex)
+        {
+            if (ex is UnauthorizedAccessException)
+            {
+                HandleUnauthorizedException();
+                return;
+            }
+
+            HandleProblem();
+
+        }
+
+        private void HandleProblem()
+        {
+            _countConsecutiveSamplingFailures++;
+
+            if (_countConsecutiveSamplingFailures >= MaxConsecutiveFailuresBeforeDisable)
+            {
+                LogMessage(LogLevel.Warn, $"After {MaxConsecutiveFailuresBeforeDisable} failed attempts, GC Metrics sampling will be disabled.");
+                Stop();
+            }
+        }
+
+        public override void Sample()
+        {
+            var currentPerfCounterInstName = default(string);
+            try
+            {
+                currentPerfCounterInstName = _pcProxyFactory.GetCurrentProcessInstanceNameForCategory(GCPerfCounterCategoryName);
```

Comment:
If I'm remembering correctly, getting the instance name can be an expensive operation if there are many processes running on a system. Will this cause any problems with the Timer such that we'll have a few samplers queued up to run back to back? Or will we have a sample every couple of minutes? I'm just wondering if we should still implement some sort of caching for this value, and then only attempt to recheck it when we run into an error?

Reply:
* I don't think we can cache the value b/c you can't guarantee that the instance name you used to have isn't in use by something else.  
* I don't think that  the caller is limited to accessing other processes performance counters based on the permission level required to do this.

What if we do this.....
* When we pull the list of names, let's put our last known instance name at the top of the list to make it the first one being checked.

Reply:
Jason did some testing using 200 instances of the same application and noticed the slowdown:
1. The initial loading of the perf counters took a few seconds
2. Subsequent loading of the perf counters took was in the 100-200 ms range

The scheduler that we're using will not queue up a bunch of sampling requests.

Since we're only sampling these values once per minute we don't think this will be a problem for the agent or the application.